README.rst
=====================================
Janggu - Deep learning for Genomics
=====================================

.. start-badges

.. image:: https://readthedocs.org/projects/janggu/badge/?style=flat
    :target: https://janggu.readthedocs.io/en/latest
    :alt: Documentation Status

.. image:: https://travis-ci.org/BIMSBbioinfo/janggu.svg?branch=master
    :alt: Travis-CI Build Status
    :target: https://travis-ci.org/BIMSBbioinfo/janggu

.. image:: https://codecov.io/github/BIMSBbioinfo/janggu/coverage.svg?branch=master
    :alt: Coverage Status
    :target: https://codecov.io/github/BIMSBbioinfo/janggu

.. image:: https://badge.fury.io/py/janggu.svg
    :alt: PyPI Package latest release
    :target: https://pypi.org/project/janggu

.. image:: https://img.shields.io/pypi/l/janggu.svg?color=green
    :alt: License
    :target: https://pypi.org/project/janggu

.. image:: https://img.shields.io/pypi/pyversions/janggu.svg
    :alt: Supported Python Versions
    :target: https://pypi.org/project/janggu/

.. image:: https://pepy.tech/badge/janggu
    :alt: Downloads
    :target: https://pepy.tech/project/janggu

.. end-badges

.. image:: jangguhex.png
   :width: 40%
   :alt: Janggu logo
   :align: center

Janggu is a python package that facilitates deep learning in the context of
genomics. The package is freely available under a GPL-3.0 license.

.. image:: Janggu-visAbstract.png
   :width: 50%
   :alt: Janggu visual abstract
   :align: center


In particular, the package allows for easy access to
typical **Genomics data formats**
and **out-of-the-box evaluation** (for keras models specifically) so that you can concentrate
on designing the neural network architecture for the purpose
of quickly testing biological hypothesis.
A comprehensive documentation is available `here <https://janggu.readthedocs.io/en/latest>`_.

Hallmarks of Janggu:
---------------------

1. Janggu provides special **Genomics datasets** that allow you to access raw data in FASTA, BAM, BIGWIG, BED and GFF file format.
2. Various **normalization** procedures are supported for dealing with of the genomics dataset, including 'TPM', 'zscore' or custom normalizers.
3. Biological features can be represented in terms of higher-order sequence features, e.g. di-nucleotide based features.
4. The dataset objects are directly consumable with neural networks for example implemented using `keras <https://keras.io>`_ or using `scikit-learn <https://scikit-learn.org/stable/index.html>`_ (see src/examples in this repository).
5. Numpy format output of a keras model can be converted to represent genomic coverage tracks, which allows exporting the predictions as BIGWIG files and visualization of genome browser-like plots.
6. Genomic datasets can be stored in various ways, including as numpy array, sparse dataset or in hdf5 format.
7. Caching of Genomic datasets avoids time consuming preprocessing steps and facilitates fast reloading.
8. Janggu provides a wrapper for `keras <https://keras.io>`_ models with built-in logging functionality and automatized result evaluation.
9. Janggu supports input feature importance attribution using the integrated gradients method and variant effect prediction assessment.
10. Janggu provides a utilities such as keras layer for scanning both DNA strands for motif occurrences.

Getting started
----------------

Janggu makes it easy to access data from genomic file formats and utilize it for
machine learning purposes.

.. code-block:: python

  dna = Bioseq.create_from_genome('dna', refgenome=<refgenome.fa>, roi=<roi.bed>)
  labels = Cover.create_from_bed('labels', bedfiles=<labels.bed>, roi=<roi.bed>)

  kerasmodel.fit(dna, labels)
  
A range of examples can be found in './src/examples' of this repository,
which includes jupyter notebooks that illustrate Janggu's functionality
and how it can be used with popular deep learning frameworks, including
keras, sklearn or pytorch.

Why the name Janggu?
---------------------

`Janggu <https://en.wikipedia.org/wiki/Janggu>`_ is a Korean percussion
instrument that looks like an hourglass.

Like the two ends of the instrument, the philosophy of the
Janggu package is to help with the two ends of a
deep learning application in genomics,
namely data acquisition and evaluation.



Installation
============

A list of python dependencies is defined in `setup.py`.
Additionally, `bedtools <https://bedtools.readthedocs.io/>`_ is required for `pybedtools` which `janggu` depends on.

Janggu depends on tensorflow and keras.
To install janggu with tensorflow version 1 and 2 use

::

   # to install with tensorflow==1.14 and keras==2.2
   pip install janggu[tf] # or janggu[tf_gpu] 

   # to install with tensorflow==2.2 and keras==2.4.3
   pip install janggu[tf2] # or janggu[tf2_gpu] 


Depending on the pip version (e.g. 20.2.2),
some package dependencies may fail to be resolved
accurately such that incompatible package versions are installed.
If this is the case, you could try using
`pip install ... --use-feature=2020-resolver`
or install the required package version manually.

Alternatively, you can install tensorflow and keras via
the conda environment using

::

   # tensorflow v1
   conda install tensorflow==1.14 keras==2.2  # or tensorflow-gpu

   # tensorflow v2
   conda install tensorflow==2.2 keras==2.4.3  # or tensorflow-gpu

Further information regarding the installation of tensorflow can be found on
the official `tensorflow webpage <https://www.tensorflow.org>`_

To verify that the installation works try to run the example contained in the
janggu package as follows

::

   git clone https://github.com/BIMSBbioinfo/janggu
   cd janggu
   python ./src/examples/classify_fasta.py single

A model is then trained to predict the class labels of two sets of toy sequencesby scanning the forward strand for sequence patterns and using an ordinary mono-nucleotide one-hot sequence encoding.
The entire training process takes a few minutes on CPU backend.
Eventually, some example prediction scores are shown for Oct4 and Mafk sequences. The accuracy should be around 85% and individual example prediction scores should tend to be higher for Oct4 than for Mafk.

You may also try to rerun the training by evaluating sequences features on both
strands and using higher-order sequence encoding using i.e. the command-line arguments: `dnaconv -order 2`.
Accuracies and prediction scores for the individual example sequences should improve compared to the previous example.

Citation
========

| Kopp, W., Monti, R., Tamburrini, A., Ohler, U., Akalin, A. Deep learning for genomics using Janggu. Nat Commun 11, 3488 (2020). https://doi.org/10.1038/s41467-020-17155-y


CHANGELOG.rst

Changelog
=========
0.10.2 (2021-09-29)
-------------------

- Fixed typos in documentation

0.10.1 (2021-06-28)
-------------------

- adapted pysam dependency
- fixed bug in coverage loading from BAM files

0.10.0 (2020-10-01)
-------------------

- added predict_variant_effect as stand-alone function.
- adapted predict_variant_effect to accept a genome in fasta format in addition to the previously required Bioseq object.
- added idx argument input_attribution to allow the selection of the i^th sequence of a Bioseq object.


0.9.9 (2020-08-04)
-------------------

- Added compatibility and tests for tensorflow 2.2.
- Adapted setup.py to enable the use with tensorflow 1.14 and tensorflow 2.2 and compatible keras versions, respectively.
- Expanded mode option for Cover.create_from_bed. New options: 'bedgraph', 'score_category' (synonym for 'categorical'), 'name_category'
- Fixed mode='categorical'/'score_category' behaviour for Cover.create_from_bed
- Mode 'bedgraph' enables compatibility with bedgraph format.
- Fixed issue with intervals overhanging the chromosome ends and intervals on non-existing chromosomes for Cover.create_from_bigwig and Cover.create_from_bam.
- Fixed issue in building DnaConv2D wrapper if the forward and revcomp layers were already built.

0.9.8 (2020-07-13)
------------------

- Improved JangguSequence which can be used with keras and tensorflow 2.0

0.9.7 (2020-05-07)
------------------

- Performance improvement for loading BED files.
- If no binsize is supplied, the binsize is detected automatically as the longest interval in the roi. Previously automatic detection was only offered if all intervals were of equal length, otherwise, a binsize needed to be specified manually.

0.9.6 (2020-02-26)
------------------

- Retired support for python 2.7
- RandomShift wrapper for data augmentation applied to DNA/Protein sequences was added by (thanks to @remomomo).
- Bedgraph files can be read using Cover.create_from_bed
- Improved efficiency of Cover.export_to_bigwig
- Improved efficiency of Cover.create_from_bed
- Internal refactoring

0.9.5 (2019-10-17)
------------------

- Variant effect prediction: added annotation argument which enables strand-specific variant effect predictions using the strandedness of features in the annotation file.
- Variant effect prediction: added ignore_reference_match argument which enables ignores mismatching nucleotides between the VCF reference base and the reference genome. By default, variant effects are only evaluated if the nucleotides agree in the reference genome and the VCF file.
- Added file validity check
- Added option to control verbosity 
- Improved efficiency for reading BAM and BIGWIG files
- Create a new cachefile with random_state only for not storing the whole genome
- Relaxed constraint for using resolution > 1 with ROI intervals. Still the interval starts have to be divisible by the resolution. Otherwise, weird rounding errors might occur.
- Fixed issue due to different numbers of network output layers.
- Added seperate dataversion to better control when cache files need to be reloaded from scratch.

0.9.4 (2019-07-15)
------------------

- Added SqueezeDim wrapper for compatibility with sklearn
- Added Transpose wrapper, replaces channel_last option of the datasets
- Loading paired-end bam-files with pairedend='5pend' option counts both ends now.
- resolution option added to create_from_array
- Relaxed restriction for sequence feature order
- Cover access via interval now returns nucleotide-resolution data regardless of the store_whole_genome option to ensure consistency.
- Refactoring


0.9.3 (2019-07-08)
------------------

- View mechanism added which allows to reuse the same dataset for different purposes, e.g. training set and test set.
- Added a dataset randomization which allows to internally randomize the data in order to avoid having to use shuffle=True with the fit method. This allows fetch randomized data in coherent chunks from hdf5 format files which improves access time.
- Added lazy loading mechanism for DNA and BED files, which defer the determination of the genome size to the dataset creation phase, but does not perform it when loading cached files to improve reload time.
- Caching logic improved in order to maximize the amount of reusability of dataset. For example, when the whole genome is loaded, the data can later be reloaded with different binsizes.
- Variant effect prediction functionality added.
- Improved efficiency for loading coverage from an array.
- Added axis option to ReduceDim
- Added Track classes to improve flexibility on plotGenomeTrack

0.9.2 (2019-05-04)
------------------

- Bugfix: Bioseq caching mechanism fixed.

0.9.1 (2019-05-03)
------------------

- Removed HTSeq dependence in favour of pybedtools for parsing BED, GFF, etc. This also introduces the requirement to have bedtools installed on the system, but it allows to parse BED-like files faster and more conveniently.
- Internal rearrangements for GenomicArray store_whole_genome=False. Now the data is stored as one array in a dict-like handle with the dummy key 'data' rather than storing the data in a fragmented fashion using as key-values the genomic interval and the respective coverages associated with them. This makes storage and processing more efficient.
- Bugfix: added conditions property to wrapper datasets.

0.9.0 (2019-03-20)
------------------

Added various features and bug fixes:

Changes in janggu.data

- Added new dataset wrapper to remove NaNs: NanToNumConverter
- Added new dataset wrappers for data augmentation: RandomOrientation, RandomSignalScale
- Adapted ReduceDim wrapper: added aggregator argument
- plotGenomeTrack added figsize option
- plotGenomeTrack added other plot types, including heatmap and seqplot.
- plotGenomeTrack refactoring of internal code
- Bioseq bugfix: Fixed issue for reverse complementing N's in the sequence.

src/janggu/cli.py
"""
Module that contains the command line app.

Why does this file exist, and why not put this in __main__?

  You may be tempted to import things from __main__ later, but that will cause
  problems: the code will get executed twice:

  - When you run `python -mjanggu` python will execute
    ``__main__.py`` as a script. That means there won't be any
    ``janggu.__main__`` in ``sys.modules``.
  - When you import __main__ it will get executed again (as a module) because
    there's no ``janggu.__main__`` in ``sys.modules``.

  Also see (1) from http://click.pocoo.org/5/setuptools/#setuptools-integration
"""

import argparse
import base64
import glob
import os
import warnings

import pandas as pd
from scipy.stats import zscore

try:
    import dash  # pylint: disable=import-error
    import dash_core_components as dcc  # pylint: disable=import-error
    import dash_html_components as html  # pylint: disable=import-error
    import plotly.graph_objs as go  # pylint: disable=import-error
except ImportError as exception:
    print('\n\ndash not available. Please install dash, dash_renderer, '
          'dash_core_components'
          ' and dash_html_components to be able to use the janggu app.\n\n')
    exit()
try:
    from sklearn.decomposition import PCA  # pylint: disable=import-error
    from sklearn.manifold import TSNE  # pylint: disable=import-error
except ImportError as exception:
    print('\n\nscikit-learn not available. Please install scikit-learn.\n\n')
    exit()

PARSER = argparse.ArgumentParser(description='janggu model explorer.')
PARSER.add_argument('-path', dest='janggu_results',
                    default=os.path.join(os.path.expanduser("~"),
                                         'janggu_results'),
                    help="Janggu results path.")

PARSER.add_argument('-port', dest='port', type=int,
                    default=8050,
                    help="Webserver port.")

ARGS = PARSER.parse_args()

APP = dash.Dash('Janggu')
APP.title = 'Janggu'
APP.config['suppress_callback_exceptions'] = True


def _serve_layer():
    return html.Div([
        dcc.Location(id='url', refresh=False),
        html.Nav(
            html.Div(
                [html.Div([
                    dcc.Link('Janggu',
                             href='/', className='navbar-brand')],
                          className='navbar-header'),
                 html.Div(
                     html.Ul([html.Li(dcc.Link('Logs', href='/logs')),
                              html.Li(dcc.Link('Model Comparison',
                                               href='/model_comparison'))],
                             className='nav navbar-nav'),
                     className='container-fluid')],
                className='navbar navbar-default navbar-expand-lg navbar-dark')),
        html.Br(),
        html.Div(id='page-content')
    ], className='container')


APP.layout = _serve_layer()


@APP.callback(dash.dependencies.Output('page-content', 'children'),
              [dash.dependencies.Input('url', 'pathname')])
def _display_page(pathname):  # pylint: disable=too-many-return-statements
    if pathname is None:
        return html.Div([])
    elif pathname == '/':
        files = glob.glob(os.path.join(ARGS.janggu_results, 'models', '*.png'))
        if not files:
            return html.Div([
                html.P('The directory "{}" appears to be empty.'
                       .format(ARGS.janggu_results))])
        encoding = {os.path.basename(os.path.splitext(name)[0]):
                    base64.b64encode(open(name, 'rb').read())
                    for name in files}

        return html.Div([
            html.Table(
                [html.Tr([html.Th('Model name'), html.Th('Architecture')])] +
                [html.Tr([dcc.Link(html.Td(name), href='/{}'.format(name)),
                          html.Td(html.Img(
                              id=name, width='50%',
                              src='data:image/png;base64,{}'.format(
                                  encoding[name].decode())))])
                 for name in encoding],
                className='table table-bordered')])
    elif pathname == '/logs':
        return html.Div(
            html.Pre(open(os.path.join(ARGS.janggu_results, 'logs', 'janggu.log')).read())
        )
    elif pathname == '/model_comparison':
        return _model_comparison_page()

    files = []
    root = os.path.join(ARGS.janggu_results, 'evaluation', pathname[1:])
    pathlen = len(root) + 1
    for root, dirnames, filenames in os.walk(root):
        for filename in filenames:
            if filename.endswith('.png') or \
                    filename.endswith('.tsv') or filename.endswith('.ply'):

                path = tuple(dirnames) + (filename,)

                files += [os.path.join(root, *path)]

    if not files:
        return html.Div([html.H3('No figures available for {}'.format(pathname[1:]))])

    image = os.path.join(ARGS.janggu_results, 'models',
                         pathname[1:] + '.png')
    encoding = base64.b64encode(open(image, 'rb').read())

    return html.Div([html.H3('Model: {}'.format(pathname[1:])),
                     html.Div([html.Div([
                         dcc.Dropdown(id='tag-selection',
                                      options=[{'label': f[pathlen:],
                                                'value': f} for f in files],
                                      value=files[0]),
                         html.Img(
                             width='100%',
                             src='data:image/png;base64,{}'.format(
                                 encoding.decode()))
                     ], className="three columns"),
                               html.Div([html.Div(id='output-plot')],
                                        className="nine columns")],
                              className='row')])


def _get_resulttables_by_name():
    combined_tables = {}
    #
    root = os.path.join(ARGS.janggu_results, 'evaluation')

    for folder, _, filenames in os.walk(root):

        mname = folder[(len(root)+1):].split('/')[0]

        for filename in filenames:

            if filename.endswith('.tsv'):
                df_ = pd.read_csv(os.path.join(folder, filename),
                                  sep='\t', header=[0], nrows=2)
                if df_.shape[0] > 1:
                    continue

                # tag + filename
                scorename = os.path.splitext(os.path.join(folder,
                                                          filename)[(len(root)
                                                                     + 2 +
                                                                     len(mname)):])[0]

                if scorename not in combined_tables:
                    combined_tables[scorename] = []

                combined_tables[scorename].append(os.path.join(folder, filename))

    return combined_tables

MODEL_COMPARISON_TABLES = _get_resulttables_by_name()

def _model_comparison_page():
    combined_tables = MODEL_COMPARISON_TABLES

    first_score = list(combined_tables.keys())[0]
    return html.Div([html.H3('Model Comparison'),
                     html.Div([html.Div([html.Label('Score'),
                                         dcc.Dropdown(id='score-selection',
                                                      options=[{'label': f,
                                                                'value': f}
                                                               for f in combined_tables],
                                                      value=first_score),
                                         html.Label('Sort'),
                                         dcc.RadioItems(id='sort-selection',
                                                        options=[
                                                            {'label': 'Ascending',
                                                             'value': True},
                                                            {'label': 'Descending',
                                                             'value': False}],
                                                        value=True),
                                         html.Label('Filter'),
                                         dcc.Input(id='filter-selection', type='text')
                                        ],
                                        className='three columns'),
                               html.Div(id='output-modelcomparison',
                                        className='nine columns')], className='row')])


@APP.callback(
    dash.dependencies.Output('output-modelcomparison', 'children'),
    [dash.dependencies.Input('score-selection', 'value'),
     dash.dependencies.Input('sort-selection', 'value'),
     dash.dependencies.Input('filter-selection', 'value')
    ])
def _update_modelcomparison(label, sorting, filterstring):

    if label is None:
        return html.P('No results for model comparison selected or detected.')

    combined_tables = MODEL_COMPARISON_TABLES
    results = combined_tables[label]
    header = ['Model', 'Layer', 'Condition', label]
    thead = [html.Tr([html.Th(h) for h in header])]
    tbody = []
    allresults = pd.DataFrame([], columns=header)

    for tab in results:
        df_ = pd.read_csv(tab, sep='\t', header=[0])
        names = df_.columns[0].split('-')
        mname, lname, cname = '-'.join(names[:-2]), names[-2], names[-1]
        if filterstring is not None and filterstring not in mname \
           and filterstring not in lname \
           and filterstring not in cname:
            continue
        allresults = allresults.append({'Model': mname,
                                        'Layer': lname,
                                        'Condition': cname,
                                        label: df_[df_.columns[0]][0]},
                                       ignore_index=True)
    allresults.sort_values(label, ascending=sorting, inplace=True)

    tbody = [html.Tr([
        dcc.Link(html.Td(allresults.iloc[i]['Model']),
                 href='/{}'.format(allresults.iloc[i]['Model']))
    ] + [
        html.Td(allresults.iloc[i][col]) for col in header[1:]
    ]) for i in range(len(allresults))]

    return html.Table(thead + tbody)


@APP.callback(
    dash.dependencies.Output('output-plot', 'children'),
    [dash.dependencies.Input('tag-selection', 'value')])
def _update_output(value):

    if value.endswith('png'):
        # display the png images directly
        img = base64.b64encode(open(value, 'rb').read())
        return html.Img(width='100%',
                        src='data:image/png;base64,{}'.format(img.decode()))

    elif value.endswith('tsv'):
        max_rows = 20
        df_ = pd.read_csv(value, sep='\t', header=[0])
        if df_.shape[0] == 1:
            df_ = df_.transpose()
            df_.columns = ['score']
        return html.Table(
            # Header
            [html.Tr([html.Th(col) for col in df_.columns])] +

            # Body
            [html.Tr([
                html.Td(df_.iloc[i][col]) for col in df_.columns
            ]) for i in range(min(len(df_), max_rows))]
        )

    elif value.endswith('ply'):
        dfheader = pd.read_csv(value, sep='\t', header=[0], nrows=0)
        # read the annotation for the dropdown
        annot = []
        for col in dfheader.columns:
            if col[:len('annot.')] == 'annot.':
                annot.append(col[len('annot.'):])

        return html.Div([
            html.Div([
                dcc.Dropdown(id='xaxis',
                             options=[{'label': x,
                                       'value': x}
                                      for x in ['Component {}'.format(i)
                                                for i in [1, 2, 3]]],
                             value='Component 1'),
                dcc.Dropdown(id='yaxis',
                             options=[{'label': x,
                                       'value': x}
                                      for x in ['Component {}'.format(i)
                                                for i in [1, 2, 3]]],
                             value='Component 2'),
                dcc.Graph(id='scatter')],
                     style={'width': '100%',
                            'display': 'inline-block',
                            'padding': '0 20'}),
            html.Div([
                dcc.Dropdown(id='operation',
                             options=[{'label': x,
                                       'value': x}
                                      for x in ['pca', 'tsne']],
                             value='pca'),
                dcc.Dropdown(id='annotation',
                             options=[{'label': x,
                                       'value': x} for x in ['None'] + annot],
                             value='None'),
                dcc.Graph(id='features')],
                     style={'width': '100%',
                            'display': 'inline-block',
                            'padding': '0 20'}),
        ], style={'columnCount': 2})

    # else the value is not known
    return html.P('Cannot find action for {}'.format(value))


@APP.callback(
    dash.dependencies.Output('scatter', 'figure'),
    [dash.dependencies.Input('tag-selection', 'value'),
     dash.dependencies.Input('operation', 'value'),
     dash.dependencies.Input('xaxis', 'value'),
     dash.dependencies.Input('yaxis', 'value'),
     dash.dependencies.Input('annotation', 'value')])
def _update_scatter(filename, operation,  # pylint: disable=too-many-locals
                    xaxis_label,
                    yaxis_label,
                    annotation):
    data = pd.read_csv(filename, sep='\t', header=[0])

    # if 'annot' in data use coloring
    colors = pd.Series(['blue'] * data.shape[0])
    for col in data:
        if col[:len('annot.')] == 'annot.':
            if annotation == col[len('annot.'):]:
                colors = data[col]
            data.pop(col)
        if col == 'row_names':
            text = data.pop(col)

    if operation == 'pca':
        pca = PCA(n_components=3)
        trdata = pca.fit_transform(data)
    else:
        tsne = TSNE(n_components=3)
        trdata = tsne.fit_transform(data)
    trdata = pd.DataFrame(trdata, columns=['Component {}'.format(i)
                                           for i in [1, 2, 3]])

    data = []
    for color in colors.unique():
        data.append(
            go.Scatter(
                x=trdata[xaxis_label][colors == color],
                y=trdata[yaxis_label][colors == color],
                text=text,
                name=color,
                customdata=colors[colors == color].index,
                mode='markers',
                marker=dict(size=3, opacity=.5)))

    return {'data': data,
            'layout': go.Layout(
                margin={'l': 40, 'b': 30, 't': 10, 'r': 0},
                xaxis={'title': xaxis_label},
                yaxis={'title': yaxis_label},
                hovermode='closest',
                plot_bgcolor='#E5E5E5',
                paper_bgcolor='#E5E5E5'
            )
           }


@APP.callback(
    dash.dependencies.Output('features', 'figure'),
    [dash.dependencies.Input('tag-selection', 'value'),
     dash.dependencies.Input('operation', 'value'),
     dash.dependencies.Input('scatter', 'selectedData')])
def _update_features(value, feature, selected):
    data = pd.read_csv(value, sep='\t', header=[0])
    data.reindex(sorted(data.columns), axis=1)

    print('_update_features')

    for col in data:
        if col[:len('annot')] == 'annot':
            data.pop(col)
        if col == 'row_names':
            data.pop(col)

    data = data.apply(zscore)
    if selected is not None:
        selected_idx = [datum['customdata'] for datum in selected['points']]
        data = data.iloc[selected_idx, :]

    mean = data.mean().values
    std = data.std().values

    return {'data': [go.Scatter(
        x=list(range(len(mean))),
        y=mean,
        mode='line',
        text=data.columns,
        error_y=dict(
            type='percent',
            value=std,
            thickness=1,
            width=0,
            color='#444',
            opacity=0.8
        ))],
            'layout': go.Layout(
                xaxis={'title': 'Features'},
                plot_bgcolor='#E5E5E5',
                paper_bgcolor='#E5E5E5',
                yaxis={'showgrid': False,
                       'title': 'Activities (z-score)'}
            )
           }


CSSES = ["https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css",
         "https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css",
         "https://codepen.io/chriddyp/pen/bWLwgP.css"]

for css in CSSES:
    APP.css.append_css({"external_url": css})

JSES = ["http://code.jquery.com/jquery-3.3.1.min.js",
        "https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"]

for js in JSES:
    APP.scripts.append_script({"external_url": js})


def main():
    print('Welcome to janggu (GPL-v3). Copyright (C) 2017-2018 '
          + 'Wolfgang Kopp.')
    warnings.DeprecationWarning("The janggu web server app will be removed in the future release.")
    APP.run_server(port=ARGS.port)
